File: euler-539-odd-elimination.pl/euler-539-odd-elimination.pl
compute(1, 1).
compute(N, Result) :-
    N > 1,
    N mod 2 =:= 0,
    N1 is N - 1,
    p(N, P),
    compute(N1, C),
    Result is (C + P) mod 987654321.

compute(N, Result) :-
    N > 1,
    N mod 2 == 0,
    M is (N - 1) // 2,
    compute(M, C),
    Result is (1 + 2 * M * (M + 1) + 4 * M - 4 * C) mod 987654321.

p(1, 1).
p(2, 2).
p(N, Result) :-
    N > 2,
    V is N // 2,
    p(V, PV),
    Result is 2 * V + 2 - 2 * PV.

main :-
    N is 1000,
    compute(N, Result),
    write(Result).

File: euler-539-odd-elimination.hs/euler-539-odd-elimination.hs
compute :: Integer -> Integer
compute 1 = 1
compute n
  | even n    = (compute (n - 1) + p n) `mod` modVal
  | otherwise = (1 + 2 * m * (m + 1) + 4 * m - 4 * compute m) `mod` modVal
  where
    modVal = 987654321
    m = (n - 1) `div` 2

p :: Integer -> Integer
p 1 = 1
p 2 = 2
p n = 2 * v + 2 - 2 * p v
  where
    v = n `div` 2

main :: IO ()
main = do
  let n = 1000000000000000000
  print $ compute n
